### ISSUE TITLE
Writes are blocked for many seconds when Scylla is not able to keep up with flushing memtables

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#1879]
*Installation details*
Scylla version (or git commit hash): master
Cluster: 1 node, -c1, -m8G
OS: F23
Disks: 1x NVMe SSD

Stress command: 
````
cassandra-stress user duration=1m cl=ONE profile=~/src/scylla-dtest/test_data/c-s-profiles/cassandra-stress-custom-large-partition-1.yaml ops\(insert=1\) -rate threads=2 -pop seq=1..100
````

Grafana dashboard:

[dirty accounting bug.pdf](https://github.com/scylladb/scylla/files/608766/dirty.accounting.bug.pdf)

You can see that occasionally we experience periods of zero throughput which lasts between 6-9 seconds and causes requests to time out.

From the analysis of the dashboard, I reached to the following conclusions.

The commit log is not responsible for the stall. Some requests block when they enter it, but the queue doesn't build up which mens they're released soon. Requests are stalled while being blocked on dirty memory, we form a long queue there.

We are constantly flushing. Usually virtual dirty (yellow line) is being released gradually and requests proceed fine. When the period of no throughput starts, virtual dirty suddenly rises over the throttling threshold of 2G by about 400M. This is directly responsible for the stall, because as long as it is above 2G, no requests will be released. When this happens, real dirty (green line) stays at the same level. This suggests that it's just a flush_reader dying at this point, reverting cleaned up memory. This is also when the next memtable will start flushing (since 5d067eebf29b1e62ae142f1c199123d736765d47). But the region is not removed from the group yet at this point, it will be when memtable region is started being merged into cache, which happens after sstable is sealed. This is also the time when commitlog segments are released and you can see on the dashboards that it happens many seconds later.

So I think we have two problems here:
  (1) We should start merging memtable into cache at the same time when we revert virtual dirty
  (2) Sealing is very slow. Maybe the fact that we flush concurrently with it has something to do with it. Need to investigate more.

**Comment by tgrabiec** [comment]
I started to look at moving cache update closer to the flush_reader dying ot see if it helps

**Comment by glommer** [comment]
@tgrabiec I have a patch for that , I am about to send.
